sh.c

```diff
@@ -1897,6 +1897,22 @@ void comp_statement(ast node, int else_if) {
     loop_nesting_level -= 1;
 
     append_glo_decl(wrap_str_lit("done"));
+  } else if (op == DO_KW) {
+    append_glo_decl(wrap_str_lit("while"));
+    loop_nesting_level += 1;
+    nest_level += 1;
+    
+    if (get_child(node, 0) != 0) {
+      comp_statement(get_child(node, 0), false);
+    } else {
+      append_glo_decl(wrap_char(':'));
+    }
+
+    append_glo_decl(comp_rvalue(get_child(node, 1), RVALUE_CTX_TEST));
+
+    nest_level -= 1;
+    loop_nesting_level -= 1;
+    append_glo_decl(wrap_str_lit("do :; done"));
```

Comment:
> creating a temporary variable that would force the first iteration of the loop

That was my first idea; but I wasn't sure if the usage of a temporary was acceptable. It makes the readability a bit harder.

> If I'm reading the code correctly, your solution generates something like this: ...

This is exactly correct; 

>  In that same thread, another solution is presented which I find...

I'll try that out; it might also solve the `set -e` limitation

Reply:
> That was my first idea; but I wasn't sure if the usage of a temporary was acceptable. It makes the readability a bit harder.

Temporaries are used for nested function calls since shells don't have something equivalent. For example, this 1-line C function:

```C
#define wrap_char(c) (-c)

text global_var(ast ident_tok) {
  return string_concat(wrap_char('_'), wrap_str_pool(get_val(ident_tok)));
}
```

is compiled to:

```shell
: $((__t1 = ident_tok = 0))
_global_var() { let ident_tok $2
  let __t1 # !Notice that the temporary __t1 is treated as a local var!
  _get_val __t1 $ident_tok
  _wrap_str_pool __t1 $__t1
  _string_concat $1 $((-(__UNDERSCORE__))) $__t1
  endlet $1 __t1 ident_tok
}
```

They can be generated with the `fresh_ident` function but they are not free in terms of performance and, as you note, generally reduce readability.